File: typescript/Behavioral/22-Visitor/Visitor/PdfExportVisitor.ts
import type { IJobVisitor } from "./IJobVisitor";
import type { PlumbingJob } from "../Element/PlumbingJob";
import type { ElectricalJob } from "../Element/ElectricalJob";
import type { CarpentryJob } from "../Element/CarpentryJob";

export class PdfExportVisitor implements IJobVisitor {
  private lines: string[] = [];

  visitPlumbing(job: PlumbingJob): void {
    this.lines.push(
      `  [PLUMBING]   #${job.jobId} | ${job.title.padEnd(32)} | ` +
        `${job.pipeMetres}m pipe @ ${job.pressureBar} bar | N$${job.cost.toFixed(2)}`,
    );
  }

  visitElectrical(job: ElectricalJob): void {
    const coc = job.requiresCertificate ? " | CoC required" : "";

    this.lines.push(
      `  [ELECTRICAL] #${job.jobId} | ${job.title.padEnd(32)} | ` +
        `${job.circuitLoadAmps}A load${coc} | N$${job.cost.toFixed(2)}`,
    );
  }

  visitCarpentry(job: CarpentryJob): void {
    this.lines.push(
      `  [CARPENTRY]  #${job.jobId} | ${job.title.padEnd(32)} | ` +
        `${job.timberMetres}m timber, ${job.finishType} | N$${job.cost.toFixed(2)}`,
    );
  }

  render(): void {
    console.log("\n=== PDF Export Preview ===");

    for (const line of this.lines) console.log(line);
  }
}

File: csharp/Behavioral/22-Visitor/Visitor/PdfExportVisitor.cs
namespace Visitor;

public class PdfExportVisitor : IJobVisitor
{
    private readonly List<string> _lines = new();

    public void Visit(PlumbingJob job) =>
        _lines.Add($"  [PLUMBING]   #{job.JobId} | {job.Title,-32} | " +
                   $"{job.PipeMetres}m pipe @ {job.PressureBar} bar | N${job.Cost:N2}");

    public void Visit(ElectricalJob job) =>
        _lines.Add($"  [ELECTRICAL] #{job.JobId} | {job.Title,-32} | " +
                   $"{job.CircuitLoadAmps}A load" +
                   $"{(job.RequiresCertificate ? " | CoC required" : "")} | N${job.Cost:N2}");

    public void Visit(CarpentryJob job) =>
        _lines.Add($"  [CARPENTRY]  #{job.JobId} | {job.Title,-32} | " +
                   $"{job.TimberMetres}m timber, {job.FinishType} | N${job.Cost:N2}");

    public void Render()
    {
        Console.WriteLine("\n=== PDF Export Preview ===");

        foreach (var line in _lines)
            Console.WriteLine(line);
    }
}
